lib/write-to-stream.ts

```diff
@@ -4,20 +4,28 @@ import { AnyIterable } from './types'
 interface IWritable {
   once: any
   write: any
+  removeListener: any
 }
 
-function waitForDrain(stream: IWritable) {
+function once(event: string, stream: IWritable): Promise<any> {
   return new Promise(resolve => {
-    stream.once('drain', resolve)
+    stream.once(event, resolve)
   })
 }
 
 async function _writeToStream(stream: IWritable, iterable: AnyIterable<any>) {
+  let errorListener
+  const errorPromise = new Promise((resolve, reject) => {
+    errorListener = reject
+    stream.once('error', reject)
+  })
   for await (const value of iterable) {
     if (stream.write(value) === false) {
-      await waitForDrain(stream)
+      await Promise.race([errorPromise, once('drain', stream)])
```

Comment:
With this, your only chance of catching an error in on the `drain` event. Considering that we don't really know or control the `highWaterMark` it could be while before we actually stop reading from the iterator.